fix: Normalize `RLIMIT_NOFILE` (LimitNOFILE) to sensible defaults

polarathene opened the pull request:
- relates to https://github.com/moby/moby/issues/38814
- fixes https://github.com/moby/moby/issues/43361
- fixes https://github.com/moby/moby/issues/45436
- fixes https://github.com/moby/moby/issues/45380


## What I did

Adjusted init configs to be as consistent as possible with handling the `RLIMIT_NOFILE` setting (_number of open file descriptors a single process can have at one time_).
- An explicit soft limit + constrained hard limit prevents a variety of issues that occur when software is run in a container vs directly on the host.
  - In the past, a soft limit may have been problematic for `dockerd` itself (_non-issue since Go 1.19 / Aug 2022).
  - `LimitNOFILE=infinity` is a regression introduced in a May 2021 PR merging a 2018 commit from another repo, whereas the value of `infinity` became excessive on various systems starting from 2019.
  - `524288` is a standard hard limit on systems with systemd since v240, being sufficient for software and not known to cause any problems.
- **NOTE:** The OpenRC config uses a single command for `ulimit` via `rc_ulimit`, preventing setting a soft and hard limit, thus it will maintain a high soft limit. You cannot use `-H` + `-S` in the same `ulimit` command to set separate soft/hard values.

<details>
<summary>Reasoning for `1024:524288` (click to expand)</summary>

- `infinity` is especially problematic as a soft limit, introducing difficult to troubleshoot / discover bugs.
  - This became more problematic with [systemd v240](https://github.com/containerd/containerd/pull/7566#issuecomment-1461140261) & [Go 1.19](https://github.com/containerd/containerd/issues/8249#issuecomment-1463400448) releases adjusting the impact implicitly.
- `1048576`  was adjusted to `524288` elsewhere to align with `docker.service` / systemd.
  - The few that actually need this to be higher [can use drop-in overrides](https://github.com/moby/moby/issues/45436#issuecomment-1534520189). However that is likely more relevant for `containerd.service` than it is `docker.service`.
- I have [documented how `LimitNOFILE` in `docker.service` + `containerd.service` affects resource usage and operations](https://github.com/moby/moby/issues/38814#issuecomment-1483951375):
  - `524288` should be more than enough to support 65k `busybox` containers (_which itself would [encounter various other bottlenecks](https://github.com/moby/moby/issues/44973#issuecomment-1543733757) along the way that require configuration changes to support_).
  - Individual containers can request higher limits with `--ulimit "nofile=soft:hard"` (_as `root` no greater than `fs.nr_open`_).
- Extensive reasoning for choosing `1024:524288` [covered here](https://github.com/containerd/containerd/pull/7566#issuecomment-1461142737).
- [systemd docs for `LimitNOFILE`](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Process%20Properties):
   > **`LimitNOFILE`:** Don't use. Be careful when raising the soft limit above 1024, since `select(2)` cannot function with file descriptors above 1023 on Linux.
   > Nowadays, the hard limit defaults to 524288, a very high value compared to historical defaults
   > 
   > Typically applications should increase their soft limit to the hard limit on their own, if they are OK with working with file descriptors above 1023, (i.e. do not use `select(2)`).
   > Note that file descriptors are nowadays accounted like any other form of memory, thus there should not be any need to lower the hard limit.

</details>

## How I did it

<details>
<summary>Approach (click to expand):</summary>

- Used `git blame` history on config to lookup relevant PRs on Github, review the changes between them and discussions.
- Engaged in main discussions on `moby` and `containerd` regarding the issue.
- Researched external resources on the subject and investigated behaviour across several releases and distros where changes from Docker, systemd and Go, or a distro affected the behaviour. Shared findings.
- Consolidated all of the information gathered to propose a positive improvement with evidence to support it.
</details>

<details>
<summary>Related History (click to expand):</summary>

- March 2014 - [Originally introduced into systemd `docker.service`](https://github.com/moby/moby/pull/4455#issuecomment-36679884) (_`LimitNOFILE` + `LimitNPROC`_):
  > _there's nothing special about `1048576`._
  > _In fact it's higher than is actually necessary (at the moment)._
- Follow-up PRs:
  - [Upstart](https://github.com/moby/moby/pull/4749) (_separate soft value introduced: `524288:1048576`_)
  - [OpenRC + SysVinit](https://github.com/moby/moby/pull/6814) (_later [SysVinit added a conditional between `-u` / `-p` based on shell](https://github.com/moby/moby/pull/7018)_)
- July 2016 - [Changed `docker.service` limits to `infinity`](https://github.com/moby/moby/pull/24307) (_[this future comment](https://github.com/moby/moby/pull/24555#issuecomment-232254840) seems to be about [this comment/issue](https://github.com/moby/moby/issues/21485#issuecomment-231376834) as the motivation for the **"non-zero limit"** config note_)
- July 2016 - https://github.com/moby/moby/pull/24555
  - `-u` + `-p` changed to `unlimited` in other init configs.
  - `LimitNOFILE` reverted back to `1048576` and relocated above the "non-zero limit" config note.
  - "non-zero limit" config note added to other init configs.
- April 2019 - [OpenRC improvements + adopts `rc_ulimit` for setting limits](https://github.com/moby/moby/pull/39000)
- May 2021 - [`LimitNOFILE` changed back to `infinity` and moved back under the "non-zero limit" config note](https://github.com/moby/moby/pull/42373)
  - This was a regression introducing problems on distros shipping systemd v240+ which redefined what `LimitNOFILE=infinity` resolves to. Some distros like Debian explicitly opt-out of that, keeping the original `1048576` limit.
  - The relocation of the line, caused confusion with users and maintainers/contributors since as it was implied that `LimitNOFILE` was set to `unlimited` for the same reason as the others (_this was originally the intent, but later realized as not applicable_).

More extensive history [coverage here](https://github.com/containerd/containerd/pull/7566#issuecomment-1461135751). `containerd` also had [a similar `LimitNOFILE` config dance](https://github.com/containerd/containerd/pull/7566#issuecomment-1452394753).

In addition, the ecosystem around Docker also changed. Notably changes introduced that affected what `infinity` resolves to (since systemd v240 in 2018Q4), and ignoring any configured soft limit (_Go 1.19/1.20 releases between Aug 2022 and May 2023_).

---

</details>

<details>
<summary>Additional references (click to expand):</summary>

- Related [change for `containerd` (review discussion)](https://github.com/containerd/containerd/pull/7566/files#r1148455486)
- [I did an extensive investigation](https://github.com/containerd/containerd/pull/7566#issuecomment-1461134171)
- Related insights on the topic I contributed:
  - https://github.com/docker/for-linux/issues/73#issuecomment-1455386927
  - https://github.com/moby/moby/issues/38814#issuecomment-1463188394

</details>

## How to verify it

Requires at least Moby 23.0.6, or 24.0.0 for an important fix introduced since Go 1.19.9 / 1.20.4:

```dockerfile
FROM alpine
RUN echo "Soft: $(ulimit -Sn)" >> /limits.txt
RUN echo "Hard: $(ulimit -Hn)" >> /limits.txt
```

```console
# `docker.service` limit applies to `docker build` since the adoption of BuildKit:
$ docker build --no-cache -t limits-test .
$ docker run --rm limits-test cat /limits.txt

Soft: 1024
Hard: 524288
```

<details>
<summary>Notes (click to expand)</summary>

- **Without this change**, both Soft and Hard values would be the same and either `1048576` (`2^20`) or `1073741816` (`2^30`).
- `grep 'files' /proc/$(pidof dockerd)/limits` will report the soft limit as the hard limit.
  
  This is expected because Go implicitly raises the soft limit since Go 1.19 released in Aug 2022. That is reset to the expected soft limit when the daemon runs some processes involved in `docker build`.
- Testing at runtime with containers using `docker run` instead of via image builds requires a related change in `containerd` (_min version required: 1.6.21 / 1.7.1_). That can be verified [as detailed here](https://github.com/containerd/containerd/pull/7566/files#r1189356383).
- Another [reproduction on limits that may be of interest](https://github.com/moby/moby/issues/38814#issuecomment-1483951375).

</details>

## Description for the changelog

Adjusted `RLIMIT_NOFILE` to more sensible defaults.


## A picture of a cute animal (not mandatory but encouraged)

![zootopia](https://i.pinimg.com/originals/22/72/7f/22727f2f26bd4217d191df9f10d6cee6.gif)

corhere commented:
Any reason not to drop the custom NOFILE limit configuration entirely and simply use the system's default, at least for the systemd unit?

polarathene commented:
@corhere If you're sure that all supported distros will be compatible with doing so, removing the line should be ok :+1: 

Alternatively, older supported distros, could be advised to use a [drop-in override](https://github.com/moby/moby/issues/45436#issuecomment-1534520189)?

**NOTE:** `contrib/init/sysvinit-redhat` had no limit configured. Not sure if that was intentional, or if it's only on systems where the hard limit is high enough? It seemed [ignored from this PR](https://github.com/moby/moby/pull/6814). Potentially no users with that init had workloads of hundred of containers. Especially today, most are probably on systemd.

---

## Concern - May be preferable to keep `LimitNOFILE` for now

I don't think this `LimitNOFILE` line would cause any problems/surprises:
- It's mostly redundant (_since it should be the default on systemd v240+_), but is somewhat informative to those not familiar with it, and would be consistent with the other supported init configs adjusting this limit.
- There may be some value in `git blame` history, which is less obvious to track when a diff for a file is only deleting lines.

I wouldn't expect the default to change in future. Nor are we likely to have someone attempt to contribute `LimitNOFILE=infinity` again as I think maintainers are well aware of this issue now.

However users may find advice to set `LimitNOFILE`, as no doubt historical search results probably land on some problems / discussions that advise changing it to some value (_usually without a soft value, as that wasn't compatible in the past_), or existing users that already have a habit to modify/set it? The new comments in this PR may help better inform those users to keep the soft limit (avoiding some potential bugs).

**Maybe keep the line around for a year or so before dropping it?:**
- Users / online resources have some time to be exposed to the change (_vs missing line and adding their own "just to be safe"_)
- Older distros with versions of systemd older than 240 may no longer need to be supported by `moby`?

---

## Concern - Dropping `LimitNOFILE` for supported distros with systemd <240

A concern that was raised back in March was needing to still support CentOS 7, which I responded with a link to a [Sep 2022 article](https://www.sobyte.net/post/2022-09/systemd-update/):
> _In `issues/23109`, systemd explicitly states that **CentOS 7 is too old** and may not be upgraded to a higher version of systemd._

Prior to systemd v240, the default hard limit apparently varied by distro. I [explored this a bit](https://github.com/containerd/containerd/pull/7566#issuecomment-1461134737), if no hard limit is present for the service, it may be as low as `4096` (kernel default). This was the case on an old Ubuntu server VM I had access to.

Originally `LimitNOFILE` was added in `docker.service` to [raise the soft limit high enough for the daemon to support managing hundreds of containers](https://github.com/containerd/containerd/pull/7566#issuecomment-1461135751):
- Since Aug 2022 (_and fix released in May 2023_) Go now handles that concern implicitly. Only the hard limit needs to be high enough now.
- Since the `LimitNOFILE` introduction in 2014, I think `containerd` later took on responsibilities that `dockerd` managed, and there was a comment saying this would make the original problem less likely? (_for `docker.service`_) Might have been about a similar concern with max tasks/threads actually.

polarathene commented:
A follow-up PR for this will apparently be needed for [`docker/docker-ce-packaging`](https://github.com/docker/docker-ce-packaging/blob/dcebf9d2dcfd61e22b7a83b442697221084c91d4/deb/common/docker-ce.docker.upstart#L6) (_I only noticed due to [this comment](https://github.com/moby/moby/pull/45548#issuecomment-1550479488)_). Only seems necessary for the upstart config?

tianon commented:
https://github.com/moby/moby/pull/45551 :eyes: (let's please not waste any cycles on Upstart :sweat_smile:)

polarathene commented on contrib/init/upstart/docker.conf:
```diff
@@ -3,7 +3,7 @@ description "Docker daemon"
 stop on runlevel [!2345]
 
-limit nofile 524288 1048576
+limit nofile 1024 524288
```
Reverting as config is being dropped.

```suggestion
limit nofile 524288 1048576
```

neersighted commented:
Could you please rebase to drop the original upstart change instead of reverting it in a later commit?

thaJeztah commented:
> Any reason not to drop the custom NOFILE limit configuration entirely and simply use the system's default, at least for the systemd unit?

Discussing in our sync call; perhaps this is the way we should go, and use systemd's defaults, which are the same currently. Does that work  for you, @polarathene ?

polarathene commented:
> use systemd's defaults, which are the same currently. Does that work for you, @polarathene ?

**That would be fine for me**, but you mentioned in the past you still support CentOS 7 (EOL in roughly a year).

That does not have a new enough systemd version and IIRC cannot support v240+, thus you'd probably fallback to `1024:4096`. That may be too low for some operations (_perhaps `prune` with enough items to remove_).

---

I also expanded on that [concern and another earlier](https://github.com/moby/moby/pull/45534#issuecomment-1550593693), click to expand the collapsed details if that's helpful.

**My stance remains that it is probably be wise to go with what this PR proposes**, and then at a later release remove the `LimitNOFILE` line, at least if there are any unexpected regressions with this change it'll be better communicated with the explicit value, before transitioning to implicit where it may differ.

neersighted commented:
I'm in favor of `LimitNOFILE` with a `TODO` to remove after systemd <= 240 (EL 7) falls away; I think it's past time to get this in to a release, and I'll try to make sure we get this reviewed this week.

cpuguy83 commented:
Its probably easier to remove it now and add a patch for older distros at packaging time.

polarathene commented:
> Its probably easier to remove it now and add a patch for older distros at packaging time.

What is the advantage of keeping a separate patch around to maintain?

Explicit defaults ensures all distros have the sane values (_that would be implicit for modern distros_), while documenting the change as a bonus (_easier to track via `git blame` too_).

When the older distros are considered unsupported, you'd remove the explicit config like you would the separate patch. This approach would transition more smoothly IMO.

It also aligns with the `openrc` and `sysvinit` configs also being adjusted here, and provides an inline description / diff for better context/visibility of the change. As I've mentioned earlier in the discussion, this has value in better informing a casual Docker user that has read third-party advice to change this setting either in future or as a past habit when applying it again in a new install of Docker.

---

I believe **keeping the value explicit** for now (_as this PR proposes_) **is more beneficial to everyone**.

neersighted commented:
Given this is both repository is both "upstream" of packagers (Docker Inc., Mirantis, Microsoft) and this is a `contrib/` file, I suggest we just do whatever we collectively think is easiest upstream.

@corhere do you have a preference? I know Mirantis maintains their own systemd units, so maybe there are no strong opinions?

polarathene commented:
> Meanwhile the default works well.

I'm fine with either outcome personally.

I would like to draw attention to the [benefits I raised in my previous comment](https://github.com/moby/moby/pull/45534#issuecomment-1616499984) of keeping the `LimitNOFILE` explicit. I don't see any disadvantage there, it's simpler than a potential patch, consistent and may be better for users to encounter.

None of that matters for me, but might be worth taking into consideration.

---

Would you like me to make any changes to this PR?

Systemd aside, this PR also normalizes all the other init configs to the same hard limit `524288` (_and the only other one that seems able to set a separate soft limit back to `1024`_). Are those changes still wanted?

cpuguy83 reviewed (changes requested):
Discussed this with maintainers.
We'd like to get this change in for 25.0 release and agreed to just remove it completely.
In packaging we'll patch the unit file as needed (pretty much just centos7).

Are you good to make those changes?

polarathene commented:
> Are you good to make those changes?

I'm not familiar with managing patches as you've proposed sorry 😅 

I'm surprised that the cited benefits of keeping the proposed changes as-is were not valued 🤔 

---

> _Are those changes still wanted?_

No feedback from your response was received about this portion of the PR.

neersighted commented:
Sorry, the idea is to drop the changes in favor of removing `LimitNOFILE` entirely. You would not have to manage the patching -- as a downstream packager, I am happy to maintain patches in Docker CE re-adding `LimitNOFILE` for systemd versions with bad defaults, and Mirantis is fine with maintaining them in their own packaging as well.

polarathene commented on contrib/init/systemd/docker.service:
```diff
@@ -26,9 +26,13 @@ StartLimitBurst=3
 # this option work for either version of systemd.
 StartLimitInterval=60s
 
+# Ensure a high enough limit for the daemon, while keeping a sane soft limit for processes.
+# The daemon will implicitly raise the soft limit to the hard limit (since Go 1.19),
+# restoring the soft limit to the original value for child processes.
+LimitNOFILE=1024:524288
+
```
Dropping `LimitNOFILE` as requested

```suggestion
```

polarathene committed:
fix: Normalize `RLIMIT_NOFILE` to sensible defaults

During review, it was decided to remove `LimitNOFILE` from `docker.service` to rely on the systemd v240 implicit default of `1024:524288`. On supported platforms with systemd prior to v240, packagers will patch the service with an explicit `LimitNOFILE=1024:524288`.

- `1024` soft limit is an implicit default, avoiding unexpected breakage. Software that needs a higher limit should request to raise the soft limit for its process.
- `524288` hard limit is an implicit default since systemd v240 and is adequate for most processes (_half of the historical limit from `fs.nr_open` of `1048576`_), while 4096 is the implicit default from the kernel (often too low). Individual containers can be started with `--ulimit` when a larger hard limit is required.
- The hard limit may not exceed `fs.nr_open` (_which a value of `infinity` will resolve to_). On most systems with systemd v240 or newer, this will resolve to an excessive size of 2^30 (over 1 billion).
- When set to `infinity` (usually as the soft limit) software may experience significantly increased resource usage, resulting in a performance regression or runtime failures that are difficult to troubleshoot.
- OpenRC current config approach lacks support for different soft/hard limits being set as it adjusts additional limits and `ulimit` does not support mixed usage of `-H` + `-S`. A soft limit of `524288` is not ideal, but 2^19 is much less overhead than 2^30, whilst a hard limit of 4096 would be problematic for Docker.

Signed-off-by: Brennan Kinney <5098581+polarathene@users.noreply.github.com>

polarathene commented:
Let me know if you'd like the two commits squashed (if you don't use squash merge), or if you'd like the history rewritten so that the `LimitNOFILE` removal is split out from the other init config changes.

Otherwise, thanks for taking the time to review and approve! ❤️

---

I've pinged the author of the equivalent containerd PR [requesting the same `LimitNOFILE` change](https://github.com/containerd/containerd/pull/7566#pullrequestreview-1544490050). If they're unresponsive, I assume a maintainer can apply that (_or I can open a fresh PR_).

neersighted reviewed (approved):
We don't squash-merge in this repository, so if you can rewrite history into a single commit, that would be appreciated. I don't think there's a need to split changes as systemd vs non-systemd.

cpuguy83 commented:
LGTM just needs a squash.

polarathene commented:
Friendly ping, not sure if you got notified of the commit from last month.

I've slightly revised the commit message today and opened an equivalent PR for `containerd` here: https://github.com/containerd/containerd/pull/8924
